feat(router): add "paramsInheritanceStrategy" router configuration option

voithos opened the pull request:
Previously, the router would merge path and matrix params, as well as
data/resolve, with special rules (only merging down when the route has
an empty path, or is component-less). This change adds an extra option
"alwaysInheritParams" which makes child routes unconditionally inherit
params from parent routes.

Closes #20572.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #20572


## What is the new behavior?
Adds an `alwaysInheritParams` extra option when configuring the router, which changes parameter inheritance to always merge from parent routes.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```


## Other information

voithos committed:
feat(router): add "paramsInheritanceStrategy" router configuration option

Previously, the router would merge path and matrix params, as well as
data/resolve, with special rules (only merging down when the route has
an empty path, or is component-less). This change adds an extra option
"paramsInheritanceStrategy" which, when set to 'always', makes child
routes unconditionally inherit params from parent routes.

Closes #20572.

klinki commented:
Woooooow, hell just froze down! I've been waiting for this for very very long time. I hope it will work as I'd expect, but if it does, then it is time for :champagne:

montella1507 commented:
@klinki  me too :D finally we have a hope, the angular router will work correctly one day.. this fix could be 100 days earlier if there were not "it works correctly" opinions of some developers though...

jasonaden commented on packages/router/src/router_module.ts:
```diff
@@ -256,6 +256,11 @@ export interface ExtraOptions {
+  /**
+   * Enables unconditional inheritance of parent params, data, and resolve.
+   */
+  alwaysInheritParams?: boolean;
```
I would like to see some other options here. Rather than boolean, let's assume the `undefined` case is default and use a string `always` for the new case. You can also have `default` which would be the same as `undefined`. Then the option can be renamed:

```
export interface ExtraOptions {
  // other stuff...

  inheritParams: 'default' | 'always';
}
```

This is more descriptive, plus gives the option to make changes down the road by adding other merge strategies.

voithos replied:
That's a good idea - done!

jasonaden reviewed (changes requested):
This is really good overall and is definitely something that will help the community. I made a few comments about changing around the naming a bit and changing the type interface slightly.

Also, when you're done you should run:

`gulp public-api:update` and add the result to your commit.

voithos commented on packages/router/src/router_state.ts:
```diff
@@ -182,28 +183,52 @@ export type Inherited = {
-    } else if (!parent.component) {
-      inhertingStartingFrom--;
+/** @internal */
+export const alwaysInheritParamsInheritorFunction: InheritorFunction =
```
It was used from a test, but it probably just belongs in said test. Done.

voithos commented on packages/router/src/router.ts:
```diff
@@ -255,11 +256,13 @@ export class Router {
-      private rootContexts: ChildrenOutletContexts, private location: Location, injector: Injector,
-      loader: NgModuleFactoryLoader, compiler: Compiler, public config: Routes) {
+      private rootContexts: ChildrenOutletContexts, private location: Location,
+      alwaysInheritParams: boolean, injector: Injector, loader: NgModuleFactoryLoader,
```
Thanks, I was wondering why the router class itself was considered part of the public api. That makes sense. To make it simpler, I decided to switch this to just be a new public property on the router (similar to `onSameUrlNavigation`), which avoids the need to add a new parameter.

voithos commented:
Thanks for the feedback!

voithos commented:
Thanks for the simplifications! I squashed the commit, and made a few alterations:
- Renamed the `InheritParamDataResolve` union type to `InheritanceStrategy`, to align better with the variables and other "strategy" namings
- Moved `InheritanceStrategy` from `router_module.ts` to `router_state.ts` so it could be shared in all usage sites, instead of defining `'default' | 'always'` multiple times
- Exported the type from `index.ts` (let me know if we shouldn't do this, but seemed reasonable since the other 'strategy' types are exported)
- Removed the `|| 'default'` bit from `router_module.ts`, since truthiness is already being checked by the `if`

jasonaden reviewed (approved):
Looks good! Let's get this in!

jasonaden commented:
You will need to run `gulp public-api:update` as this will add `InheritanceStrategy` to the public API.

jasonaden commented:
You should be able to ignore the failing Travis build. That one is a flake.

IgorMinar commented on packages/router/src/router_state.ts:
```diff
@@ -174,36 +175,59 @@ export class ActivatedRoute {
+ *
+ * @stable
+ */
+export type InheritanceStrategy = 'default' | 'always';
```
can you call it something else than `default`? if we change the default to `always` in the future then the `default` will be deprecated and non-default.

voithos replied:
Good point, done. I wasn't sure how best to encode the (somewhat odd) default behavior. Currently have it as `emptyonly`, as in, "only inherit params from parents for empty [pathless, componentless] routes".

Definitely open to other suggestions.

IgorMinar commented on tools/public_api_guard/router/router.d.ts:
```diff
@@ -327,6 +331,7 @@ export declare class Router {
     config: Routes;
     errorHandler: ErrorHandler;
     readonly events: Observable<Event>;
+    inheritanceStrategy: InheritanceStrategy;
```
`inheritanceStrategy` name is too vague. the name makes no reference to `params` which is what is actually being configured. I suggest `paramsInheritanceStrategy` instead.

voithos replied:
Agreed, done.

IgorMinar commented on tools/public_api_guard/router/router.d.ts:
```diff
@@ -157,6 +158,9 @@ export declare class GuardsCheckStart extends RouterEvent {
 }
 
+/** @stable */
+export declare type InheritanceStrategy = 'default' | 'always';
```
I'd prefer this to be inlined in the `Router` definition because it just pollutes the namespace and we did that for `onSameUrlNavigation` which is also string union type.

voithos replied:
That would mean that we'd have to maintain the set `'default' | 'always'` in at least 2 places (`onSameUrlNavigation` already has this problem; it's defined in both router.ts and router_module.ts).

Are you sure that's the cleaner approach?

IgorMinar commented on packages/router/src/router_state.ts:
```diff
@@ -174,36 +175,59 @@ export class ActivatedRoute {
 
+/**
+ * @whatItDoes Defines how the router merges params, data and resolved data from parent to child
+ * routes. Set this option to 'always' to enable unconditional inheritance of parent params, data,
```
can you reformat this as a list with two options. it will be much easier to read that way.

voithos replied:
Done.

IgorMinar reviewed (changes requested):
we need to weak the api a bit before we roll this out.

mary-poppins commented:
You can preview b5cbe32 at https://pr20678-b5cbe32.ngbuilds.io/.
You can preview 12cb4a3 at https://pr20678-12cb4a3.ngbuilds.io/.
You can preview 3d1283d at https://pr20678-3d1283d.ngbuilds.io/.
You can preview dc46bd0 at https://pr20678-dc46bd0.ngbuilds.io/.
You can preview 994a95d at https://pr20678-994a95d.ngbuilds.io/.

IgorMinar commented on tools/public_api_guard/router/router.d.ts:
```diff
@@ -237,6 +238,9 @@ export declare type Params = {
 };
 
+/** @stable */
+export declare type ParamsInheritanceStrategy = 'emptyonly' | 'always';
```
`emptyonly` sounds good, but can you camelcase it please? we use camelcase for string literals. example: https://pr20678-994a95d.ngbuilds.io/api/router/RunGuardsAndResolvers

voithos replied:
Sure, done.

IgorMinar commented on tools/public_api_guard/router/router.d.ts:
```diff
@@ -127,6 +127,7 @@ export interface ExtraOptions {
     errorHandler?: ErrorHandler;
     initialNavigation?: InitialNavigation;
     onSameUrlNavigation?: 'reload' | 'ignore';
+    paramsInheritanceStrategy?: ParamsInheritanceStrategy;
```
and yeah. I'd rather inline the type in the two places for now to make the api docs less confusing.

we have a general problem with type alias where we don't have a good way to document them and display the docs, so it's easier if the docs are inlined that way autocompletion and angular.io displays the docs in a better way. (and yes, we have cases where we haven't inlined the types already, but that was an oversight)

voithos replied:
Alright, thanks for the explanation. I went ahead and inlined it at the router site as well as in router_module.ts, but it's actually used in ~6 other locations as well. It felt a little weird copying this around so much, so I tried keeping the type alias, marking it as `@internal`, and removing it from the exports. That way it shouldn't show up in non-public APIs, but the internal stuff can still use a centralized alias. What do you think?

Let me know if I should just inline it everywhere instead.

IgorMinar reviewed (commented):
other than these two changes, this PR looks good. thanks!

IgorMinar commented:
also if you rebase your PR against master, the flake from e2e should disappear. (I hope!)

mary-poppins commented:
You can preview 577fac2 at https://pr20678-577fac2.ngbuilds.io/.

voithos commented on packages/router/testing/src/router_testing_module.ts:
```diff
@@ -84,12 +84,15 @@ export class SpyNgModuleFactoryLoader implements NgModuleFactoryLoader {
     urlSerializer: UrlSerializer, contexts: ChildrenOutletContexts, location: Location,
     loader: NgModuleFactoryLoader, compiler: Compiler, injector: Injector, routes: Route[][],
-    urlHandlingStrategy?: UrlHandlingStrategy) {
+    opts: ExtraOptions = {}, urlHandlingStrategy?: UrlHandlingStrategy) {
```
One thing that might be worth pointing out: this change adds `opts` as ExtraOptions *before* `urlHandlingStrategy`. I think I originally did that because it matches the order in `setupRouter()` from `router_module.ts`. But I suppose if someone was calling `setupTestingRouter` directly, the changing parameter order might break them.

Should I add it at the end instead? The downside there is that the order won't match the order in `setupRouter()`.

IgorMinar replied:
wow.. good catch. I asked @jasonaden to add @angular/router/testing to our api guard so that the api changes are highlighted in the code reviews.

And you are right. This change is a breaking change and we can't land it as is. I suggest that we change the api to: 

```
export function setupTestingRouter(
     urlSerializer: UrlSerializer, contexts: ChildrenOutletContexts, location: Location,
     loader: NgModuleFactoryLoader, compiler: Compiler, injector: Injector, routes: Route[][],
     options?: ExtraOptions | UrlHandlingStrategy) {
```

or even better as an overload with the `UrlHandlingStrategy` as the last param deprecated.

Can you please take a stab at that? thanks!

voithos replied:
Thanks for the suggestions! I tried implementing it the way you mentioned, setting up some overloads (let me know if I didn't do the `@deprecated` part right). It did also require adding a type guard, and I'm not sure if it can be implemented in a better way (normally, I'd be worried about e.g. property renaming breaking a guard like this, but I don't think anyone runs their test code through an optimizer, right?).

I also just noticed, though, that `setupRouter` and `setupTestingRouter` have diverged in other ways, unfortunately. :( For example, one has `injector` before `loader`, whereas the order is reversed in the other. Maybe all this overloading isn't worth doing, and we should just live with the mismatched ordering? Happy to go either way here.

IgorMinar replied:
setupRouter is a private api so we have the freedom to change it. setupTestingRouter is public api so that api can't change. I'm not sure how easy it is to make the two consistent but if it's a small tweak then please go ahead and make it in this PR. otherwise we should deal with it in a separate followup PR.

I'll review the overloads shortly. thanks

voithos replied:
Ok, I see. Looking more closely, the main positional difference now is that `injector` is in the wrong spot (it should be 2 args earlier, before `loader`). Moving it would require changing those 3 args (`injector`, `loader`, and `compiler`) to take union types, to handle both variants. After that, we could add a `routeReuseStrategy` to the end, which should be easy (won't require moving anything).

The only other major difference is that `setupRouter` accepts an `ApplicationRef` - I'm guessing that doesn't even exist at test time? Maybe that's intentionally missing?

I could potentially do these changes now if you feel the cleanup is useful, but I think it could also be done in a future PR without conflicting with the current deprecation/compatibility code this PR adds, as long as no new arg needs to be inserted in the middle of the signature (so we should be ok as long as `ApplicationRef` is intentionally not there).

mary-poppins commented:
You can preview 1f89e13 at https://pr20678-1f89e13.ngbuilds.io/.

IgorMinar commented on packages/router/testing/src/router_testing_module.ts:
```diff
@@ -76,6 +76,11 @@ export class SpyNgModuleFactoryLoader implements NgModuleFactoryLoader {
 
+function isUrlHandlingStrategy(opts: ExtraOptions | UrlHandlingStrategy):
+    opts is UrlHandlingStrategy {
+  return 'shouldProcessUrl' in opts;
```
how about using `instanceof UrlHandlingStrategy` instead? in which case you can just inline it below and don't need this guard at all.

voithos replied:
That won't quite work because `UrlHandlingStrategy` is an interface and won't exist at runtime, right? We could do `instanceof DefaultUrlHandlingStrategy`, but then that would only take care of the default strategy and would fail for a user-provided one.

(Added a comment to the type guard mentioning this rationale)

IgorMinar replied:
`UrlHandlingStrategy` is an abstract class: https://github.com/angular/angular/blob/bebedfed24d6fbfa492e97f071e1d1b41e411280/packages/router/src/url_handling_strategy.ts#L16

so the instanceof check should work just fine.

voithos replied:
Ah, my mistake. Still, I think my original point stands because implementations don't use `extends` with the abstract class, they use [`implements`](https://github.com/angular/angular/blob/bebedfed24d6fbfa492e97f071e1d1b41e411280/packages/router/src/url_handling_strategy.ts#L42), which doesn't actually result in any class relationship (and thus `instanceof` fails). Here's an example illustrating this, with a fake `UrlHandlingStrategy` and `DefaultUrlHandlingStrategy`: https://www.typescriptlang.org/play/#src=abstract%20class%20UrlHandlingStrategy%20%7B%0D%0A%20%20abstract%20shouldProcessUrl()%3A%20void%3B%0D%0A%20%20abstract%20extract()%3A%20void%3B%0D%0A%20%20abstract%20merge()%3A%20void%3B%0D%0A%7D%0D%0A%0D%0Aclass%20DefaultUrlHandlingStrategy%20implements%20UrlHandlingStrategy%20%7B%0D%0A%20%20shouldProcessUrl()%20%7B%20%7D%0D%0A%20%20extract()%3A%20void%20%7B%20%7D%0D%0A%20%20merge()%3A%20void%20%7B%20%7D%0D%0A%7D%0D%0A%0D%0Alet%20u%20%3D%20new%20DefaultUrlHandlingStrategy()%3B%0D%0A%0D%0Aconsole.log('instanceof%20DefaultUrlHandlingStrategy'%2C%20u%20instanceof%20DefaultUrlHandlingStrategy)%3B%0D%0Aconsole.log('instanceof%20UrlHandlingStrategy'%2C%20u%20instanceof%20UrlHandlingStrategy)%3B

Just as a sanity check, I tried adding some logging [in this conditional block](https://github.com/angular/angular/blob/bebedfed24d6fbfa492e97f071e1d1b41e411280/packages/router/testing/src/router_testing_module.ts#L91), and indeed, `urlHandlingStrategy instanceof UrlHandlingStrategy` is `false`.

IgorMinar reviewed (commented):
a few more comments, the rest looks good from the api perspective.

IgorMinar commented:
btw Jason is working on #21046 which will break this PR if he lands it before you because of the additional API guard coverage that as you pointed out was previously missing. Let's see who can resolve the review comments first :-)

voithos commented on packages/router/testing/src/router_testing_module.ts:
```diff
@@ -84,9 +89,38 @@ export class SpyNgModuleFactoryLoader implements NgModuleFactoryLoader {
+/**
+ * Router setup factory function used for testing.
+ *
+ * @deprecated The 2nd-to-last argument should be `ExtraOptions`, not `UrlHandlingStrategy`
```
Done.

voithos commented:
Thanks for the feedback! I also decided to add a really small unit test that makes sure that the `ExtraOptions` injection works properly for the deprecated ordering in `setupTestingRouter`. It's in `integration.spec.ts` - let me know if I should change anything there.

mary-poppins commented:
You can preview 746baa1 at https://pr20678-746baa1.ngbuilds.io/.

IgorMinar commented:
The new test looks good. Let's remove the typeguard as I suggested above and land this PR.

It would be great if you made a followup PR with the setupRouter signature changes to make things more consistent.

voithos commented:
I think the type guard might still be necessary (see my reply).

mary-poppins commented:
You can preview c50a178 at https://pr20678-c50a178.ngbuilds.io/.

IgorMinar commented:
This is so bizarre. I don't understand why we use `implements` there instead of `extends` looks like an oversight to me. That or `UrlHandlingStrategy` should be (have been) an interface rather than an abstract class.

I asked @jasonaden about this.

Since we can always refactor the typeguard later, let's leave it as is.

IgorMinar commented:
the CI failures look like flakes. I'm restarting the jobs.

voithos commented:
The CI failure didn't look related to this PR, and it seems that the build is failing on master? https://travis-ci.org/angular/angular/builds

IgorMinar commented:
yeah. that's unrelated. I marked the PR for merging so the caretaker
(person on rotation) will merge it and get it into g3 soon.

On Tue, Dec 19, 2017 at 9:20 AM Zaven Muradyan <notifications@github.com>
wrote:

> The CI failure didn't look related to this PR, and it seems that the build
> is failing on master? https://travis-ci.org/angular/angular/builds
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/20678#issuecomment-352826992>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AANM6K1Ghj7_yS7noheDCyGGKSQ_nrsNks5tB_BmgaJpZM4QuS8l>
> .
>

voithos commented:
Thank you!
